docs/source/design/input_output_specifications.md

```diff
@@ -0,0 +1,692 @@
+    range       : Tuple[float, float] = (0.0, 1.0)    # Expected value range for the metric
+    threshold   : Optional[float] = None              # Optional threshold for pass/fail
+    metadata    : Dict[str, Any] = {}                 # Arbitrary additional metadata
+```
+
+
+
+
+###  `TaskSpec`
+
+
+```python
+class TaskSpec(BaseModel, extra="forbid"):
+    name               : str
+    goal               : Literal[ # type
+        "clustering", "classification",
+        "dimensionality_reduction", "perturbation",
+        "spatial_registration", "custom"
+    ]                                         # Task goal/type
+    input_schema       : InputOutputSpecs   # Model input schema
+    prediction_schema  : InputOutputSpecs   # Model output schema
+    reference_schema   : InputOutputSpecs   # Ground truth schema from dataset
+    metrics            : list[MetricSpec]      # List of metrics to evaluate
+    context            : dict[str, Any] = {}   # Organism, tissue, etc.
+    metadata           : dict[str, Any] = {}   # Additional metadata
+```
+
+
+## Optionals
+
+
+### Axis, Provenance, Resource
+
+```py
+# Add and validate dimensions od datasest if applicabale
+class Axis(BaseModel, extra="forbid"):
+    name : str                         # "cells", "genes", "x", "time", …
+    size : Optional[int] = None        # None = variable
+    kind : Literal[
+        "sample", "feature", "spatial", "channel",
+        "sequence", "time", "batch", "custom"
+    ] = "custom"
+
+# Dataset Provenance. May be useful in cli for search capabilities and VCP integration
+class Provenance(BaseModel, extra="forbid"):
+    source   : Literal["internal", "external"] = "external"
+    version  : str
+    authors  : Optional[List[str]]
+    license  : Optional[str]
+    citation : Optional[str]
+    url      : Optional[HttpUrl]
+    doi      : Optional[str]
+
+
+# Useful to validate hardware resource requirement for a model
+class Resource(BaseModel, extra="forbid"):
+  use_gpu            : bool = False
+  gpu_type       : str
+  gpu_count      : int = 1                              # Minimum number of GPUs required
+  gpu_min_memory     : Optional[int] = None                 # Minimum memory per GPU in MB
+  gpu_features       : set[str] = set()                     # e.g., {"flash_attention", "fp8"}
+  gpu_mem_mb     : Optional[int] = None                 # Total GPU memory required (legacy, prefer min_mem_mb)
+  gpu_compute_capability : Optional[str] = None         # e.g., "7.0", "8.0" (NVIDIA CUDA compute capability)
+  multi_gpu      : Optional[bool] = False               # Whether multi-GPU is required
+  cpu_cores      : int = 4
+  ram_mb         : int = 8_192
+  disk_mb        : int = 10_240
+  cloud_provider : Optional[Literal["aws", "gcp", "azure", "custom"]] = None
+  instance_type  : Optional[str] = None                 # e.g., "n1-standard-8", "m5.large"
```

Comment:
`recommend_instance_types: Optional[List[Tuple(CloudProvider, InstanceType)]]` 

Also, this seems unnecessary if we have all the above cpu/gpu/mem requirements, since once can infer the correct instance type from that for a given cloud provider. But since it's optional, it seems like a nice convenience to provide this.